helix-term/src/application.rs

```diff
@@ -245,6 +256,131 @@ impl Application {
         }
     }
 
+    pub async fn handle_debugger_message(&mut self, payload: helix_dap::Payload) {
+        use helix_dap::{events, Event};
+        let mut debugger = match self.editor.debugger.as_mut() {
+            Some(debugger) => debugger,
+            None => return,
+        };
+
+        match payload {
+            Payload::Event(ev) => match ev {
+                Event::Stopped(events::Stopped {
+                    thread_id,
+                    description,
+                    text,
+                    reason,
+                    all_threads_stopped,
+                    ..
+                }) => {
+                    debugger.is_running = false;
+                    let main = debugger
+                        .threads()
+                        .await
+                        .ok()
+                        .and_then(|threads| threads.get(0).cloned());
+
+                    if let Some(main) = main {
+                        let (bt, _) = debugger.stack_trace(main.id).await.unwrap();
+                        debugger.stack_pointer = bt.get(0).cloned();
+                        debugger.stopped_thread = Some(main.id);
+                    }
+
+                    let scope = match thread_id {
+                        Some(id) => format!("Thread {}", id),
+                        None => "Target".to_owned(),
+                    };
+
+                    let mut status = format!("{} stopped because of {}", scope, reason);
+                    if let Some(desc) = description {
+                        status.push_str(&format!(" {}", desc));
+                    }
+                    if let Some(text) = text {
+                        status.push_str(&format!(" {}", text));
+                    }
+                    if all_threads_stopped.unwrap_or_default() {
+                        status.push_str(" (all threads stopped)");
+                    }
+
+                    if let Some(helix_dap::StackFrame {
+                        source:
+                            Some(helix_dap::Source {
+                                path: Some(src), ..
+                            }),
+                        line,
+                        column,
+                        end_line,
+                        end_column,
+                        ..
+                    }) = &debugger.stack_pointer
+                    {
+                        let path = src.clone();
+                        let line = *line;
+                        let column = *column;
+                        let end_line = *end_line;
+                        let end_column = *end_column;
```

Comment:
I think you can simplify this by using `debugger.stack_pointer` (no &) then using `Some(ref src)` if needed

Reply:
Thanks, just got lost while trying to refactor that :)